frame/glutton/src/lib.rs

```diff
@@ -0,0 +1,303 @@
+	/// would possibly be underestimated in that case.
+	#[pallet::storage]
+	pub(super) type TrashData<T: Config> = StorageMap<
+		Hasher = Twox64Concat,
+		Key = u32,
+		Value = [u8; 1024],
+		QueryKind = OptionQuery,
+		MaxValues = ConstU32<65_000>,
+	>;
+
+	/// Keeps track of how many entries we have inside `TrashData`.
+	#[pallet::storage]
+	pub(crate) type TrashDataCount<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn integrity_test() {
+			assert!(
+				!T::WeightInfo::waste_ref_time_iter(1).ref_time().is_zero(),
+				"Weight zero; would get stuck in an infinite loop"
+			);
+			assert!(
+				!T::WeightInfo::waste_proof_size_some(1).proof_size().is_zero(),
+				"Weight zero; would get stuck in an infinite loop"
+			);
+		}
+
+		fn on_idle(_: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+			let mut meter = WeightMeter::from_limit(remaining_weight);
+			if !meter.check_accrue(T::WeightInfo::empty_on_idle()) {
+				return T::WeightInfo::empty_on_idle()
+			}
+
+			let proof_size_limit = Storage::<T>::get().mul_floor(meter.remaining().proof_size());
+			let computation_weight_limit =
+				Compute::<T>::get().mul_floor(meter.remaining().ref_time());
+			let mut meter = WeightMeter::from_limit(Weight::from_parts(
+				computation_weight_limit,
+				proof_size_limit,
+			));
+
+			Self::waste_at_most_proof_size(&mut meter);
+			Self::waste_at_most_ref_time(&mut meter);
+
+			meter.consumed
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Initializes the pallet by writing into `TrashData`.
+		///
+		/// Only callable by Root. A good default for `trash_count` is `5_000`.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::DbWeight::get().writes((*trash_count).into()))]
+		pub fn initialize_pallet(origin: OriginFor<T>, trash_count: u32) -> DispatchResult {
+			ensure_root(origin)?;
+
+			// Fill up the `TrashData` storage item.
+			(0..trash_count).for_each(|i| TrashData::<T>::insert(i, &[i as u8; 1024]));
+			TrashDataCount::<T>::put(trash_count);
+
+			Self::deposit_event(Event::PalletInitialized);
+			Ok(())
+		}
+
+		/// Expands `TrashData` by the specified amount. Limit is 65k entries.
+		///
+		/// Only callable by Root.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::DbWeight::get().writes((*trash_count).into()))]
+		pub fn expand_trash_data(origin: OriginFor<T>, trash_count: u32) -> DispatchResult {
+			ensure_root(origin)?;
+
+			let current_trash_count = TrashDataCount::<T>::get();
+
+			(current_trash_count..trash_count)
+				.for_each(|i| TrashData::<T>::insert(i, &[i as u8; 1024]));
+			TrashDataCount::<T>::put(current_trash_count.saturating_add(trash_count));
+
+			Self::deposit_event(Event::TrashDataUpdated);
+			Ok(())
+		}
+
+		/// Shrinks `TrashData` by the specified amount.
+		///
+		/// Only callable by Root.
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::DbWeight::get().writes((*trash_count).into()))]
```

Comment:
This is not the correct weight, since it misses the proof the component.  
I will add a small bench but probably replace it with a counted map anyway.